server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -23,56 +24,131 @@
 import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.index.query.QueryRewriteContext;
 import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
 import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.common.Strings.format;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
 
 /**
  * A query that performs kNN search using Lucene's {@link org.apache.lucene.search.KnnFloatVectorQuery} or
  * {@link org.apache.lucene.search.KnnByteVectorQuery}.
  *
- * NOTE: this is an internal class and should not be used outside of core Elasticsearch code.
  */
 public class KnnVectorQueryBuilder extends AbstractQueryBuilder<KnnVectorQueryBuilder> {
     public static final String NAME = "knn";
+    private static final int NUM_CANDS_LIMIT = 10000;
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField QUERY_VECTOR_BUILDER_FIELD = new ParseField("query_vector_builder");
```

Comment:
I think we shouldn't have `query_vector_builder` in the knn query. 

It seems like we should instead of a `semanticVectorQuery` that rewrites itself to a `knn` query after fetching the correct information.

We were hamstrung by the current top level `knn` design.

Reply:
Thanks for the feedback. Are you suggesting to introduce a new query like this?

```
"semantic_search": {
    "field": "dense-vector-field",
    "num_candidates": 100,
    "query_vector_builder": {
      "text_embedding": { 
        "model_id": "my-text-embedding-model", 
        "model_text": "The opposite of blue" 
      }
    }
  }
```
What are your reasons why we can't have `query_vector_builder` inside a knn query?  There would be a lot of duplication for  `semantic_query`.

Reply:
> What are your reasons why can't we have query_vector_builder inside a knn query? There would be a lot of duplication for semantic_query.

It complicates the query. We have rewrite phases, etc. to make this sort of thing simpler. 

I think that a "semantic_search" query can simplify things and is the correct way to handle this. 

It could even dynamically handle switching between `dense` and `sparse` models.

IMO A semantic search query would be something like
```
"semantic_search": {
    "field": "dense-vector-field",
    "num_candidates": 100,
    "model_id": "my-text-embedding-model",
    "text": "The opposite of blue"
  }
```

Reply:
Thanks for further clarifying, Ben! I will remove `query_vector_builder` from knn query.